frame/evm/src/runner/stack.rs

```diff
@@ -827,6 +910,222 @@ where
+				return Ok(U256::from(actual_size));
+			};
+		}
+		Ok(U256::from(<Pallet<T>>::account_code_metadata(address).size))
+	}
+
+	#[cfg(not(feature = "evm-with-weight-limit"))]
+	fn code_size(&mut self, address: H160) -> Result<U256, ExitError> {
+		Ok(U256::from(<Pallet<T>>::account_code_metadata(address).size))
+	}
+
+	#[cfg(feature = "evm-with-weight-limit")]
+	fn code_hash(&mut self, address: H160) -> Result<H256, ExitError> {
+		let size_limit: u64 = self
+			.metadata()
+			.gasometer()
+			.config()
+			.create_contract_limit
+			.unwrap_or_default() as u64;
+		if let Some(weight_info) = self.weight_info_mut() {
+			// First try to record fixed sized `AccountCodesMetadata` read
+			// Temptatively 20 + 8 + 32
+			// TODO we need a way to check whether AccountCodesMetadata for an address is already
+			// recorded in this transaction, otherwise we are over accounting..
+			weight_info.try_record_proof_size_or_fail(ACCOUNT_CODES_METADATA_PROOF_SIZE)?;
+
+			if <AccountCodesMetadata<T>>::get(address).is_none() {
+				// If it does not exist, try to record `create_contract_limit` first.
+				weight_info.try_record_proof_size_or_fail(size_limit)?;
+				let meta = Pallet::<T>::account_code_metadata(address);
+				let actual_size = meta.size;
+				// Refund if applies
+				weight_info.refund_proof_size(size_limit.saturating_sub(actual_size));
+				return Ok(meta.hash);
+			};
+		}
+		Ok(<Pallet<T>>::account_code_metadata(address).hash)
+	}
+
+	#[cfg(not(feature = "evm-with-weight-limit"))]
+	fn code_hash(&mut self, address: H160) -> Result<H256, ExitError> {
+		Ok(<Pallet<T>>::account_code_metadata(address).hash)
+	}
+
+	#[cfg(feature = "evm-with-weight-limit")]
+	fn record_external_dynamic_opcode_cost(
+		&mut self,
+		opcode: Opcode,
+		gas_cost: GasCost,
+		target: evm::gasometer::StorageTarget,
+	) -> Result<(), ExitError> {
+		let size_limit: u64 = self
+			.metadata()
+			.gasometer()
+			.config()
+			.create_contract_limit
+			.unwrap_or_default() as u64;
+
+		let weight_info = if let Some(weight_info) = self.weight_info_mut() {
+			weight_info
+		} else {
+			return Ok(());
+		};
+
+		// Record ref_time first
+		// TODO benchmark opcodes, until this is done we do used_gas to weight conversion for ref_time
+
+		// Record proof_size
+		// Return if proof size recording is disabled
+		let proof_size_limit = if let Some(proof_size_limit) = weight_info.proof_size_limit {
+			proof_size_limit
+		} else {
+			return Ok(());
+		};
+
+		let maybe_record = match gas_cost {
+			GasCost::ExtCodeSize { target_is_cold }
+			| GasCost::Balance { target_is_cold }
+			| GasCost::ExtCodeHash { target_is_cold }
+			| GasCost::CallCode { target_is_cold, .. }
+			| GasCost::StaticCall { target_is_cold, .. }
+			| GasCost::ExtCodeCopy { target_is_cold, .. }
+			| GasCost::SLoad { target_is_cold, .. }
+			| GasCost::DelegateCall { target_is_cold, .. }
+			| GasCost::SStore { target_is_cold, .. }
+			| GasCost::Call { target_is_cold, .. } => target_is_cold,
```

Comment:
Does warm-up even make sense for substrate chains? I think it would be fine to ignore user-provided access_list on parachains for weight computation, and don't perform actual warm-up on transaction.

Reply:
What is certain is that we need a mechanism to avoid counting several of the same keys/values, because in real world popular networks like moonbeam, most of the traffic goes through a small number of "popular" contracts, so we will lose too much blockspace if we count several times the access to the same data.

On the other hand, we can't rely on the access_list, because it is a mechanism managed by transaction, and not by block.

@tgmichel we need to create our own access recorder at block level,  the problem is that there is no way in the evm to keep context beetween transactions, so the only way is to create a new storage item `Accessed` in pallet evm to keep track of accessed addresses and storage slots. Off course, this storage item should be kill in `on_finalize` to not pollute the PoV or the persisted storage.

Reply:
> we need to create our own access recorder

@librelois done, added two separate storage maps one for codes and one for storage slots